internal/datacoord/channel_manager.go

```diff
@@ -0,0 +1,280 @@
+const (
+	bgCheckInterval  = 3 * time.Second
+	maxWatchDuration = 20 * time.Second
+)
+
+// ChannelManager manages the allocation and the balance of channels between datanodes
+type ChannelManager struct {
+	ctx              context.Context
+	mu               sync.RWMutex
+	posProvider      positionProvider
+	store            RWChannelStore
+	factory          ChannelPolicyFactory
+	registerPolicy   RegisterPolicy
+	deregisterPolicy DeregisterPolicy
+	assignPolicy     ChannelAssignPolicy
+	reassignPolicy   ChannelReassignPolicy
+	bgChecker        ChannelBGChecker
+}
+
+type channel struct {
+	name         string
+	collectionID UniqueID
+}
+
+// ChannelManagerOpt is to set optional parameters in channel manager
+type ChannelManagerOpt func(c *ChannelManager)
+
+func withFactory(f ChannelPolicyFactory) ChannelManagerOpt {
+	return func(c *ChannelManager) { c.factory = f }
+}
+func defaultFactory(hash *consistent.Consistent) ChannelPolicyFactory {
+	return NewConsistentHashChannelPolicyFactory(hash)
+}
+
+// NewChannelManager return a new ChannelManager
+func NewChannelManager(ctx context.Context, kv kv.TxnKV, posProvider positionProvider, options ...ChannelManagerOpt) (*ChannelManager, error) {
+	hashring := consistent.New()
+	c := &ChannelManager{
+		ctx:         ctx,
+		posProvider: posProvider,
+		factory:     defaultFactory(hashring),
+		store:       NewChannelStore(kv),
+	}
+
+	if err := c.store.Reload(); err != nil {
+		return nil, err
+	}
+
+	for _, opt := range options {
+		opt(c)
+	}
+
+	c.registerPolicy = c.factory.NewRegisterPolicy()
+	c.deregisterPolicy = c.factory.NewDeregisterPolicy()
+	c.assignPolicy = c.factory.NewAssignPolicy()
+	c.reassignPolicy = c.factory.NewReassignPolicy()
+	c.bgChecker = c.factory.NewBgChecker()
+	return c, nil
+}
+
+// Startup adjust the channel store according to current cluster states
+func (c *ChannelManager) Startup(nodes []int64) error {
+	channels := c.store.GetNodesChannels()
+	olds := make([]int64, 0, len(channels))
+	for _, c := range channels {
+		olds = append(olds, c.NodeID)
+	}
+
+	newOnlines := c.getNewOnlines(nodes, olds)
+	for _, n := range newOnlines {
+		if err := c.AddNode(n); err != nil {
+			return err
+		}
+	}
+
+	offlines := c.getOfflines(nodes, olds)
+	for _, n := range offlines {
+		if err := c.DeleteNode(n); err != nil {
+			return err
+		}
+	}
+	log.Debug("cluster start up",
+		zap.Any("nodes", nodes),
+		zap.Any("olds", olds),
+		zap.Int64s("new onlines", newOnlines),
+		zap.Int64s("offlines", offlines))
+	return nil
+}
+
+func (c *ChannelManager) bgCheckChannelsWork(ctx context.Context) {
+	timer := time.NewTicker(bgCheckInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			c.mu.Lock()
+
+			channels := c.store.GetNodesChannels()
+			reallocs, err := c.bgChecker(channels)
+			if err != nil {
+				log.Warn("channel manager bg check failed", zap.Error(err))
+
+				c.mu.Unlock()
+				continue
+			}
+
+			updates := c.reassignPolicy(c.store, reallocs)
```

Comment:
while we are reassigning, what if addNode or deleteNode happens?
need 	c.mu.Lock() for protection?

Reply:
already hold c.mu and unlock it after reassign